File: 백준/Silver/2468. 안전 영역/안전 영역.cc
#include <bits/stdc++.h>
using namespace std;

const int dx[] = {0, 1, 0, -1};
const int dy[] = {-1, 0 , 1, 0};
int n;
int a[102][102], visited[102][102];

void dfs(int y, int x, int z){
	visited[y][x] = 1;
	for (int i=0; i < 4; i++){
		int nx = x + dx[i];
		int ny = y + dy[i];
		if (nx < 0 || ny < 0 || nx >= n || ny >= n) continue;
		if (visited[ny][nx] == 1) continue;
		if (a[ny][nx] < z) continue;
		dfs(ny, nx, z);
	}
}

int main(){
	int max_area = -1;
	cin >> n;
	
	for(int i=0; i<n; i++){
		for(int j=0; j<n; j++){
			cin >> a[i][j];
		}
	}

	for (int i = 1; i <= 100; i++){
		int cnt = 0;
		fill(&visited[0][0], &visited[101][102], 0);
		for (int y = 0; y < n; y++){
			for (int x =0; x < n; x++){
				if (a[y][x] >= i && !visited[y][x]){
					cnt++;
					dfs(y, x, i);
				}
			}
		}
		max_area = max(max_area, cnt);
	}

	cout << max_area << "\n";
	return 0;
}

File: 백준/Silver/2468. 안전 영역/안전 영역.py
import sys

sys.setrecursionlimit(10**6)
input = sys.stdin.readline

n = int(input())
mp = [list(map(int, input().split())) for _ in range(n)]
dy = [-1, 0, 1, 0]
dx = [0, 1, 0, -1]

def dfs(y, x, height):
    visited[y][x] = 1
    
    for i in range(4):
        ny = y + dy[i]
        nx = x + dx[i]
        
        if 0 <= ny < n and 0 <= nx < n:
            if visited[ny][nx] == 0 and mp[ny][nx] > height:
                dfs(ny, nx, height)
    
mx = -1
for i in range(n):
    for j in range(n):
        mx = max(mx, mp[i][j]) 

result = 1
for i in range(1, mx + 1):
    visited = [[0] * n for _ in range(n)]
    cnt = 0
    for j in range(n):
        for k in range(n):
            if visited[j][k] == 0 and mp[j][k] > i:
                cnt += 1
                dfs(j,k, i)
    result = max(result, cnt)
    
print(result)
